### PR TITLE
Added A2CGAE

### PR DESCRIPTION
**Pull request by sriram13m**
Added Actor Critic with generalized advantage estimation.

This is a continued work of https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/pull/6

### CODE REVIEW

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+function RLBase.update!(learner::A2CGAELearner, experience)
+    AC = learner.approximator
+    γ = learner.γ
+    λ = learner.λ
+    w₁ = learner.actor_loss_weight
+    w₂ = learner.critic_loss_weight
+    w₃ = learner.entropy_loss_weight
+    states, actions, rewards, terminals, rollout, next_state = experience
+    states = send_to_device(device(AC), states)
+    rollout = reshape(rollout, : , size(rollout,2)*size(rollout,3))
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: We can use `flatten_batch` here to support states of dimensions > 1.

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+function RLBase.extract_experience(t::CircularCompactSARTSATrajectory, learner::A2CGAELearner)
+    if isfull(t)
+        (
+            states = get_trace(t, :state),
+            actions = get_trace(t, :action),
+            rewards = get_trace(t, :reward),
+            terminals = t[:terminal],
+            rollout = t[:state],
+            next_state = select_last_frame(get_trace(t, :next_state)),
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: `next_state` is not necessary here. Explained above.

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    update!(AC, gs)
+end
+
+function RLBase.extract_experience(t::CircularCompactSARTSATrajectory, learner::A2CGAELearner)
+    if isfull(t)
+        (
+            states = get_trace(t, :state),
+            actions = get_trace(t, :action),
+            rewards = get_trace(t, :reward),
+            terminals = t[:terminal],
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: Better to use `get_trace` to keep it consistant, though `get_tract(t, :terminal) === t[:terminals]` here.

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    next_state_values = AC(next_state, Val(:V))
+    gains = discount_rewards(
+        rewards,
+        γ;
+        dims = 2,
+        init = send_to_host(next_state_values),
+        terminal = terminals,
+    )
+    gains = send_to_device(device(AC), gains)
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: Here I think `gains = advantages .+= select_last_dim(rollout_values, 1:(nframes(rollout_values)-1))` would be enough

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    next_state = send_to_device(device(AC), next_state)
+
+    states_flattened = flatten_batch(states) # (state_size..., n_thread * update_step)
+    actions = flatten_batch(actions)
+
+    actions = CartesianIndex.(actions, 1:length(actions))
+
+    rollout_values = AC(rollout, Val(:V))
+    rollout_values = send_to_host(rollout_values)
+    rollout_values = reshape(rollout_values,size(states,2),size(states,3)+1)
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: Same as above, `size(states, 2)` and `size(states, 3)` may not be the value you want for states of dimension greater than 1. Replace it with size of the last two dimensions.

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        terminal = terminals,
+    )
+    gains = send_to_device(device(AC), gains)
+
+    gs = gradient(Flux.params(AC)) do
+        probs = AC(states_flattened, Val(:Q))
+        log_probs = log.(probs)
+        log_probs_select = probs[actions]
+        values = AC(states_flattened, Val(:V))
+        advantage = vec(gains) .- vec(values)
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl:68]: Obviously, here we have calculated `values` twice. The `values` should be the same with `rollout_values` except for the last frame. I'm wondering if we can only calculate it once.

**Review by findmyway — COMMENTED** [review]
Good job!

File: src/algorithms/policy_gradient/A2CGAE.jl (hunk 4 above)
Comment (sriram13m) [reply src/algorithms/policy_gradient/A2CGAE.jl]: Could you expand on why this might be true?

Comment (findmyway) [reply src/algorithms/policy_gradient/A2CGAE.jl]: Sure, based on the [`general_advantage_estimation`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/master/src/utils/base.jl#L392) you have implemented, we can calculate the GAE with Eq16 in the original [paper](https://arxiv.org/pdf/1506.02438.pdf). 

Since what we want is the (biased) returns/gains, here we can directly add the state value to the advantage part here. And then use this value to guide the critic network.

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    advantages = flatten_batch(advantages)
+    advantages = send_to_device(device(AC), advantages)
+
+    gs = gradient(Flux.params(AC)) do
+        probs = AC(states_flattened, Val(:Q))
+        log_probs = log.(probs)
+        log_probs_select = probs[actions]
+        values = AC(states_flattened, Val(:V))
+        advantage = vec(gains) .- vec(values)
+        actor_loss = -mean(log_probs_select .* Zygote.dropgrad(advantages))
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl]: `advantages` -> `advantage` ?

File: src/algorithms/policy_gradient/A2CGAE.jl [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        γ,
+        λ;
+        dims = 2,
+        terminal = terminals,
+    )
+
+    gains = advantages + select_last_dim(rollout_values, 1:(nframes(rollout_values)-1))
+    gains = send_to_device(device(AC), gains)
+    advantages = flatten_batch(advantages)
+    advantages = send_to_device(device(AC), advantages)
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2CGAE.jl:61]: I thought that `advantages` is not used anymore, am I right? If that is the case, we can reuse the allocation of `advantages` and calculate gains by `gains = advantages .+= select_last_dim(rollout_values, 1:(nframes(rollout_values)-1))`

File: src/algorithms/policy_gradient/A2C.jl [hunk 9]
```diff
@@ -95,3 +95,21 @@ function RLBase.extract_experience(t::CircularCompactSARTSATrajectory, learner::
...
+    update!(agent.policy, agent.trajectory)
+
+    # the main difference is we'd like to flush the buffer after each update!
+    if isfull(agent.trajectory)
+        empty!(agent.trajectory)
+        push!(agent.trajectory; state = state, action = action)
+    end
+
+    action
+end
```
Comment (findmyway) [inline src/algorithms/policy_gradient/A2C.jl:115]: Could you merge the latest master branch?

File: src/algorithms/policy_gradient/A2CGAE.jl (hunk 7 above)
Comment (sriram13m) [reply src/algorithms/policy_gradient/A2CGAE.jl]: Advantages is estimated from generalized advantage estimation to update the actor whereas advantage is estimated for updating the critic. So I think it should be advantages here. However Zygote.dropgrad(advantages ) is not necessary.

File: src/algorithms/policy_gradient/A2CGAE.jl (hunk 8 above)
Comment (sriram13m) [reply src/algorithms/policy_gradient/A2CGAE.jl:61]: Unfortunately advantages is used to update the actor and hence we can't reuse the allocation.

File: src/algorithms/policy_gradient/A2CGAE.jl (hunk 7 above)
Comment (findmyway) [reply src/algorithms/policy_gradient/A2CGAE.jl]: Yeah, that makes sense.

File: src/algorithms/policy_gradient/A2CGAE.jl (hunk 8 above)
Comment (findmyway) [reply src/algorithms/policy_gradient/A2CGAE.jl:61]: I see. Thanks!

File: test/A2CGAE.jl [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                    )
+                ) |> gpu,
+                optimizer=ADAM(1e-3),
+                kind=HYBRID_APPROXIMATOR
+            ),
+            γ=0.99f0,
+            λ=0.99f0,
+            actor_loss_weight = 1.0f0,
+            critic_loss_weight = 0.5f0,
+            entropy_loss_weight = - 0.001f0
```
Comment (findmyway) [inline test/A2CGAE.jl]: It should be a positive number for the `entropy_loss_weight`. Sorry that it's my fault. I've corrected it in the test case of `A2C.jl`.
